### PR TITLE
Remove warning regarding Android parallel being experimental, add robust version check

### LINKED ISSUE #2297: Lose the 'parallelism-is-experimental on Android' warning in detox logs
**Is your feature request related to a problem? Please describe.**
Can get rid of that warning now:
```
WARN:  [test.js] Multiple workers is an experimental feature on Android and requires an emulator binary of version 28.0.16 or higher. Check your version by running: $ANDROID_HOME/tools/bin/sdkmanager --list
```

**Describe the solution you'd like**
Need to:
1. Remove the warning.
2. Check emulator version on a best-effort basis, see if fits; Restrict warnings to when emulator version is either old or unknown.

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by d4vidi**
- [x] This is a small change 
- [x] This change has been discussed in issue #2297 and the solution has been agreed upon with maintainers.

---

**Description:**
Resolves #2297

### CODE REVIEW

**Comment by noomorph** [comment]
Will check today.

File: detox/src/devices/drivers/android/emulator/AVDValidator.js [hunk 1]
```diff
@@ -1,28 +1,54 @@
...
   async validate(avdName) {
     const avds = await this._avdsResolver.resolve(avdName);
+    this._assertAVDs(avds);
+    await this._assertAVDMatch(avds, avdName);
+    await this._validateEmulatorVer();
+  }
+
+  _assertAVDs(avds) {
     if (!avds) {
       const avdmanagerPath = path.join(environment.getAndroidSDKPath(), 'tools', 'bin', 'avdmanager');
```
Comment (noomorph) [inline detox/src/devices/drivers/android/emulator/AVDValidator.js:23]: Should belong to `utils/environment.js`, i.e. `environment.getAVDManagerPath()`.

File: detox/src/devices/drivers/android/emulator/AVDValidator.js [hunk 2]
```diff
@@ -1,28 +1,54 @@
...
 
       throw new Error(`Could not find any configured Android Emulator.\n
```
Comment (noomorph) [inline detox/src/devices/drivers/android/emulator/AVDValidator.js:25]: Maybe, let's convert it to `new DetoxRuntimeError({ message, hint })`, please?

File: detox/src/devices/drivers/android/emulator/AVDValidator.js [hunk 3]
```diff
@@ -1,28 +1,54 @@
...
         Try creating a device first, example: ${avdmanagerPath} create avd --force --name Pixel_API_28 --abi x86_64 --package "system-images;android-28;default;x86_64" --device "pixel"
         or go to https://developer.android.com/studio/run/managing-avds.html for details on how to create an Emulator.`);
     }
+  }
 
+  _assertAVDMatch(avds, avdName) {
     if (_.indexOf(avds, avdName) === -1) {
       throw new Error(`Can not boot Android Emulator with the name: '${avdName}',
```
Comment (noomorph) [inline detox/src/devices/drivers/android/emulator/AVDValidator.js:33]: Maybe, let's convert it to new DetoxRuntimeError({ message, hint }), please?

File: detox/src/devices/drivers/android/emulator/AVDValidator.js [hunk 4]
```diff
@@ -1,28 +1,54 @@
...
+  async _validateEmulatorVer() {
+    const emulatorVersion = await this._emulatorVersionResolver.resolve();
+    if (!emulatorVersion) {
+      logger.warn({ event: 'AVD_VALIDATION' }, 'Emulator version detection failed (See previous logs)');
+      return;
+    }
+
+    if (emulatorVersion.major < REQUIRED_EMULATOR_MAJOR) {
+      logger.warn({ event: 'AVD_VALIDATION' }, [
+          `Your installed emulator binary version (${emulatorVersion.toString()}) is too old, and may not be suitable for parallel test execution.`,
```
Comment (noomorph) [inline detox/src/devices/drivers/android/emulator/AVDValidator.js:47]: `${emulatorVersion.toString()}` is too verbose, use just `${emulatorVersion}`, why not?

File: detox/src/devices/drivers/android/emulator/EmulatorDriver.js [hunk 5]
```diff
@@ -26,11 +27,12 @@ class EmulatorDriver extends AndroidDriver {
   constructor(config) {
     super(config);
 
-    this.freeDeviceFinder = new FreeEmulatorFinder(this.adb, this.deviceRegistry);
     const emulatorExec = new EmulatorExec();
     this._emuVersionResolver = new EmulatorVersionResolver(emulatorExec);
     this._emuLauncher = new EmulatorLauncher(emulatorExec);
-    this._avdValidator = new AVDValidator(emulatorExec);
+    this._avdsResolver = new AVDsResolver(emulatorExec);
```
Comment (noomorph) [inline detox/src/devices/drivers/android/emulator/EmulatorDriver.js:33]: Was there any specific reason to pollute `this` with a never used `_avdsResolver` private property?

File: detox/src/devices/drivers/android/emulator/AVDValidator.test.js [hunk 6]
```diff
@@ -1,34 +1,48 @@
...
+    versionResolver.resolve.mockResolvedValue('');
 
-  it('should use an AVDs resolver', async () => {
-    avdsResolverObj().resolve.mockResolvedValue(['mock-avd-name']);
+    const AVDValidator = require('./AVDValidator');
+    uut = new AVDValidator(avdsResolver, versionResolver);
+  });
 
-    await uut.validate('mock-avd-name');
+  const givenExpectedAVD = () => avdsResolver.resolve.mockResolvedValue(['mock-avd-name']);
```
Comment (noomorph) [inline detox/src/devices/drivers/android/emulator/AVDValidator.test.js:21]: Wow, nice! :) Kudos for `mockResolvedValue`, אדע להבא!

**Review by noomorph — APPROVED** [review]
In general, the PR looks very good to me, although there are a few minor comments to it.
If possible, please consider them, @d4vidi.

Also, my last musing here is the following:

Since you have implemented an outdated emulator version check, I wonder, would not it be vise to put that version validation into `AndroidDriver` itself, somewhere in `allocateDevice()`... so that users of `AttachedAndroidDriver` will benefit from your changes as well?

Regards,
Yaroslav.

File: detox/src/devices/drivers/android/emulator/EmulatorDriver.js (hunk 5 above)
Comment (d4vidi) [reply detox/src/devices/drivers/android/emulator/EmulatorDriver.js:33]: Of course - will double check!

**Comment by d4vidi** [comment]
@noomorph thanks a lot for the review! I will address all proposed suggestions.
As per scope - this validation is only relevant to emulators, and hence not available for other component to use, such as real devices (i.e. through the _android_ driver).

File: detox/src/devices/drivers/android/emulator/AVDValidator.js (hunk 2 above)
Comment (d4vidi) [reply detox/src/devices/drivers/android/emulator/AVDValidator.js:25]: Have you considered introducing a lint rule for this? 🤔

**Pull request merged**